### PR TITLE
Change analyzer executor to do an intersection with the given span 

### LINKED ISSUE #27703: Possibly wrong `span1.Contains(span2)` "order"
On VSfM sometimes using `ICodeFixService` returns wrong list of code fixes if code changed recently(last ~1sec). I debugged what is happening and my conclusion is that most likely problem is at https://github.com/dotnet/roslyn/blob/9e79136/src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs#L604
What we do from VSfM code, we call `ICodeFixService.GetFixesAsync()` and we use `span` and we put cursor position as `span` so something like `(193,193)` and there is diagnostic at `(190,200)` for `CS0103`(The name XXX does not exist in current context). Which will later be used to produce `Add using ...` or `Generate XXX`...  fixes.
But this doesn't not happen because https://github.com/dotnet/roslyn/blob/9e79136/src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs#L604 filters this diagnostic out, hence no fixes. Should this `span.Value.Contains(diagnostic.Location.SourceSpan)` be `diagnostic.Location.SourceSpan.Value.Contains(span)`?

### PR DESCRIPTION
**Pull request by Therzok**
For cases where the requested span is (x, x), try using span
intersection rather than contains.

That should cover all the cases for requesting the diagnostics
at caret

Fixes #27703

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Is there a test that can be written that demonstrates the cahnge in behavior?

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs [hunk 1]
```diff
@@ -601,7 +601,7 @@ private IEnumerable<DiagnosticData> ConvertToLocalDiagnosticsWithCompilation(Doc
                         continue;
                     }
 
-                    if (span.HasValue && !span.Value.Contains(diagnostic.Location.SourceSpan))
+                    if (span.HasValue && !span.Value.IntersectsWith(diagnostic.Location.SourceSpan))
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs]: This also matches with the [test](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AnalyzerExecutor.cs,1649) used by the core analyzer driver in the compiler layer to determine if a diagnostic is a local or non-local diagnostic.

**Comment by heejaechang** [comment]
ping @Therzok ?

**Comment by Therzok** [comment]
I didn't have time to write a unit test yet, other stuff was on :fire:, will get to it tomorrow

**Comment by Therzok** [comment]
Could I get some guidance on where to add the test?

**Comment by Therzok** [comment]
Don't see anything tested at this layer, so which layer should the test be on?

**Comment by heejaechang** [comment]
@Therzok add here
http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures2.UnitTests/Diagnostics/DiagnosticServiceTests.vb,58

**Comment by heejaechang** [comment]
due to some historical reason, most of tests code for feature layer is in editorfeature layer (it used to be 1 dll and we split it to feature and editorfeature dll but we left tests in 1 dll - editorfeature unit tests)

**Comment by Therzok** [comment]
Will spin up my windows machine, this is counterproductive :p

**Comment by Therzok** [comment]
I'm wondering if I should add some offsets to the span start

**Comment by heejaechang** [comment]
that editorfeaturetest2 dll is vb only because we wanted to use vb's xml syntax stuff. except that, you should be able to test both csharp and vb in that dll

**Comment by Therzok** [comment]
Does the PR look good now? Failures look unrelated?

**Comment by heejaechang** [comment]
@Therzok this might need to move to master as well. tagging @jinujoseph in case, he want it to be in 15.9

**Comment by heejaechang** [comment]
retest windows_release_vs-integration_prtest please

**Comment by jinujoseph** [comment]
This was called as VS-Mac blocker 
@DavidKarlas @KirillOsenkov does it needs to be in 15.9 or master (16.0) is fine ?

**Comment by KirillOsenkov** [comment]
@jinujoseph as long as we can consume this fix via a set of Roslyn Nuget packages, we don't care where those packages are produced from :)

**Comment by KirillOsenkov** [comment]
@jinujoseph apologies, my mistake, Marius tells me we need this in 15.9.

**Comment by jinujoseph** [comment]
@KirillOsenkov thanks , would you help fill out the ask mode template defining the importance to take this for 15.9

**Comment by KirillOsenkov** [comment]
@jinujoseph where's the template?

**Comment by jinujoseph** [comment]
@KirillOsenkov I copied a template above to be filled

**Comment by jinujoseph** [comment]
@KirillOsenkov  ping ?

**Comment by KirillOsenkov** [comment]
@Therzok I'm on vacation this week, would you be able to fill out the template at the top of this issue?

Basically they need to know what is the customer facing implications of this, what doesn't work as expected, are there any workarounds for us, is the fix risky, etc. Thanks!

Customer scenario
Bugs this fixes
Workarounds, if any
Risk
Performance impact
Is this a regression from a previous update?
Root cause analysis
Details

**Comment by heejaechang** [comment]
ping?

**Comment by DavidKarlas** [comment]
### Customer scenario
Inside Visual Studio for Mac, when user right-clicks in text editor and requests "Quick Fix bulb" VSfM requests diagnostics/quick fixes from Roslyn just for that location when right-click was clicked, and not whole line. Which results in `.Contains()` returning empty list instead of `Intersecting` quick-fixes with that location. This problem doesn't appear always, but only if diagnostics weren't generated yet for latest version of file(sometimes it takes second or two for this to happen), so it happens only if user just recently modified file. This is most annoying when there are multiple "Add using" missing, so if user quickly clicks on 1st missing and QuickFix appears correctly and user applies fix, 2nd time user clicks on different location QuickFix menu will show 0 quick fixes, because this bug.
### Bugs this fixes
This fixes #27703
### Workarounds, if any
In theory VSfM could change to request for whole line and then filter, but then what if diagnostic spans across multiple lines, do we request for whole file?
### Risk
I can't think of any regression this could cause.
### Performance impact
None imo.
### Is this a regression from a previous update?
No.
### Root cause analysis
There are some bug reports from users reporting sometimes QuickFix comes up empty, its most likely related to this. But mostly I found it on my own seeing pattern if I request QuickFix quickly after changing code it doesn't show QuickFixes, so I debugged and pinpointed problem to this line.
### Details

**Comment by jinujoseph** [comment]
Approved to merge for 15.9.preview4 via [link](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/695450)

**Comment by heejaechang** [comment]
retest WIP please

**Comment by heejaechang** [comment]
@jasonmalinowski can you check this in. I think there is some issue with WIP monitoring stuff. 

not sure how I can re-run WIP thing. also doesn't want to close and reopen just for WIP and re-run all tests again.

**Comment by heejaechang** [comment]
@Therzok can you rewrite the history? 

> WIP test case for empty span request

if commit message has WIP, we can't check in

**Comment by jasonmalinowski** [comment]
The WIP bot is flagging that there's still a commit in the history with WIP. That'll need to be squashed or rewritten.

**Comment by KirillOsenkov** [comment]
I've squashed, removed the WIP from the commit and force-pushed to Marius' branch.

**Pull request merged**